tailcfg/tailcfg.go

```diff
@@ -1250,13 +1288,64 @@ const (
 	PeerCapabilityIngress PeerCapability = "https://tailscale.com/cap/ingress"
 )
 
+// NodeCapMap is a map of capabilities to their optional values. It is valid for
+// a capability to have no values (nil slice); such capabilities can be tested
+// for by using the HasCapability method.
+type NodeCapMap map[NodeCapability][]RawMessage
+
+// Equal reports whether c and c2 are equal.
+func (c NodeCapMap) Equal(c2 NodeCapMap) bool {
+	if len(c) != len(c2) {
+		return false
+	}
+	for k, v := range c {
+		v2, ok := c2[k]
+		if !ok {
+			return false
+		}
+		if !slices.EqualFunc(v, v2, func(e1, e2 RawMessage) bool {
+			return bytes.Equal([]byte(e1), []byte(e2))
+		}) {
+			return false
+		}
+	}
+	return true
+}
+
+// UnmarshalNodeCapJSON unmarshals each JSON value in cm[cap] as T.
+// If cap does not exist in cm, it returns (nil, nil).
+// It returns an error if the values cannot be unmarshaled into the provided type.
+func UnmarshalNodeCapJSON[T any](cm NodeCapMap, cap NodeCapability) ([]T, error) {
+	vals, ok := cm[cap]
+	if !ok {
+		return nil, nil
+	}
+	out := make([]T, 0, len(vals))
+	for _, v := range vals {
+		var t T
+		if err := json.Unmarshal([]byte(v), &t); err != nil {
+			return nil, err
+		}
+		out = append(out, t)
+	}
+	return out, nil
+}
+
+// HasCapability reports whether c has the capability cap.
+// This is used to test for the existence of a capability, especially
+// when the capability has no values.
```

Comment:
maybe reword "has no values" somehow to be more clear?

"has no associated argument/data values"
"has no ancillary values"

We should come up with a good name for that optional value bit and use its name consistently.